builder/simplify.cc

```diff
@@ -882,33 +932,64 @@ class simplifier : public node_mutator {
       return;
     } else if (const block* b = body.as<block>()) {
       scoped_trace trace("licm");
-      // Put the stmts that actually depend on the loop inside the loop.
+      // This LICM is more aggressive than the typical compiler. Because we can freely add or remove loops around calls,
+      // if we find a loop invariant stmt that consumes a loop variant stmt, we can try to force the loop varying stmt
+      // to be loop invariant, and remove both from the loop.
       std::vector<stmt> result;
       std::vector<stmt> loop_body;
-      std::set<var> produced_by_loop;
       loop_body.reserve(b->stmts.size());
-      for (const stmt& i : b->stmts) {
-        if (!depends_on(i, op->sym).any()) {
-          // This stmt does not depend on the loop variable. Find out which buffers are consumed by it.
-          std::set<var> consumed_by_i;
-          buffers_accessed_via_aliases(i, /*consumed=*/true, consumed_by_i);
-
-          if (empty_intersection(consumed_by_i, produced_by_loop)) {
-            // This doesn't depend on the loop and doesn't consume anything produced in the loop, lift it out.
-            if (i.defined()) result.push_back(i);
-          } else {
-            // This stmt consumes something produced by the loop we have so far, we can't lift it out of the loop.
-            buffers_accessed_via_aliases(i, /*consumed=*/false, produced_by_loop);
-            loop_body.push_back(i);
-          }
+
+      // Find out if i produces something consumed by a loop invariant.
+      for (auto ri = b->stmts.rbegin(); ri != b->stmts.rend(); ++ri) {
+        stmt i = *ri;
+
+        if (!depends_on(i, op->sym).var) {
+          // i is loop invariant. Add it to the lifted result.
+          result.push_back(std::move(i));
         } else {
-          // This stmt depends on the loop.
-          buffers_accessed_via_aliases(i, /*consumed=*/false, produced_by_loop);
-          loop_body.push_back(i);
+          // i depends on the loop. We are effectively reordering result ahead of i, we need to make sure we can do
```

Comment:
I'm bit worried there might be some tricky corner cases, I don't have a precise example, but something like:

loop {
A
B
C
}

let's say C is loop invariant and B is not, but we can make it loop invariant. After that the logic goes to handle `A`, but fails to do that wrt to 'C', so 'C' is removed from the 'result' list, but 'B' still might stay in 'result' even though it doesn't make sense anymore, because the only reason it was there because we wanted to lift  'C`.

Reply:
Added a test case that covers this. It works because the reordering checks all stmts it's trying to lift every time, and if it fails it puts them back in the loop.

However, I just realized, it will drop the crops in such cases, and if there are other loop invariant statements, those crops will be dropped inside the loop. This is technically correct, but adding wasted computation. I'll fix that.

Reply:
Actually, it seems to work even in this case too...

Reply:
Right, I'm worrying about a hypothetical case where we have a chain of liftable functions and there is another function which can be lifted, but only some of the items from result are returned back.